horovod/common/controller.cc

```diff
@@ -156,7 +164,8 @@ ResponseList Controller::ComputeResponseList(std::atomic_bool& shut_down,
 
   bool need_communication = true;
   if (response_cache_.capacity() > 0 &&
-      !cache_coordinator.uncached_in_queue()) {
+      !cache_coordinator.uncached_in_queue() &&
+      !cache_coordinator.just_joined()) {
```

Comment:
Join is different from other ops because it's done rank-by-rank, at different times.
All ranks need to communicate if any other rank just sent Join request, because this may be the last joined rank and Join response will be sent by the communicator.
These other ranks will not have the Join request in their own queue, so the cache is used to let them know someone else just joined.

Reply:
Thanks for the explanation! I still don't think the additional status is necessary though. If a worker has a Join request in their queue, it will set the `uncached_in_queue()` status flag since the Join requests are not cached. This will trigger the communication path with the coordinator when any worker has a Join request in their queue. If it is the last joined rank, the `uncached_in_queue()` flag will be set, the coordinator communication will trigger, and the Join response should be created/sent to all workers. This seems to result in the same behavior as the `just_joined()` flag.

Reply:
I think you're correct. But I need to verify/implement "set the uncached_in_queue() status flag" part - I don't think it works for Join requests now. This should be easy and doesn't require additional flags.